design/upgrade.md

```diff
@@ -0,0 +1,165 @@
+While migrations are possible, they are certainly not desirable since they require extra upgrade logic to be written and tested, as well as providing new potential paths for failure.
+Going forward, it will be important for the Rook project to increase our discipline regarding the introduction of breaking changes.
+We should be **very** careful about adding any new code that requires a migration during the update process.
+
+### **Kubernetes Built-in Support**
+Kubernetes has some [built-in support for rolling updates](https://kubernetes.io/docs/tasks/run-application/rolling-update-replication-controller/) with the `kubectl rolling-update` command.
+Rook can potentially take advantage of this support for our replication controllers that have multiple stateless pods deployed, such as RGW.
+This support is likely not a good fit for some of the more critical and sensitive components in the cluster, such as monitors, that require careful observation to ensure health is maintained and quorum is reestablished.
+
+If the upgrade controller uses the built-in rolling update support for certain stateless components, it should still verify all cluster health checks before proceeding with the next set of components.
+
+### **Synchronization**
+The upgrade process should be carefully orchestrated in a controlled manner to ensure reliability and success.
+Therefore, there should be some locking or synchronization that can ensure that while an upgrade is in progress, other changes to the cluster cannot be made.
+For example, if the upgrade controller is currently rolling out a new version, it should be be possible to modify the cluster CRD with other changes, such as removing a node from the cluster.
+
+This upgrade locking is also relevant to how the upgrade controller remembers to resume the upgrade process after it has updated its own version and restarted itself to begin running the new version and resume where it left off.
+
+### **Scalability**
+For small clusters, the process of upgrading one pod at a time should be sufficient.
+However, for much later clusters (100+ nodes), this would result in an unacceptably long upgrade window duration.
+The upgrade controller should be able to batch some of its efforts to upgrade multiple pods at once in order to finish an upgrade in a more timely manner.
+
+This batching should not be done across component types (e.g. upgrading mons and OSDs at the same time), those boundaries where the health of the entire cluster is verified should still exist.
+But, the upgrade controller should be able to slowly increase its component update batch size as it proceeds through each component type.
+For example, in true canary deployment fashion, a single monitor could be upgraded to the new version and mon health/quorum will be verified.
+Then two monitors could be updated at once and verification occurs again, followed by four monitors, etc. up to a reasonable upper bound.
```

Comment:
I would think we always update the mons one at a time, despite the cluster size. Even for large clusters you wouldn't have more than a handful of mons. And there are typically the same number of mgrs as the mons. So really we're just talking about batching the osds and possibly rgw and mds.

Reply:
good point, monitors should probably only be upgraded one at a time.